presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PredicatePushDown.java

```diff
@@ -837,6 +845,93 @@ public PlanNode visitSemiJoin(SemiJoinNode node, RewriteContext<Expression> cont
             return output;
         }
 
+        private PlanNode visitFilteringSemiJoin(SemiJoinNode node, RewriteContext<Expression> context)
+        {
+            Expression inheritedPredicate = context.get();
```

Comment:
I think you can reuse: `PredicatePushDown.Rewriter#processInnerJoin` method. Then any conjuct from: `InnerJoinPushDownResult#getJoinPredicate` can be used as filter in semi join (all should be equivalent).

Reply:
I tried that but it gets awkward:
* you have to pass weird `boolean allowNonDeterministicPushDownToLeftSide` parameter to `processInnerJoin`
* you have to split `InnerJoinPushDownResult.getJoinPredicate` into actual join predicate and post-join predicate to be put within `FilterNode` above `SemiJoinNode`. The `InnerJoinPushDownResult.postJoinPredicate` is always TRUE as InnerJoin handles non-equality filter expressions internally.

IMO it does not get more readable. And you do not save many lines of code.

I captured the work in progress in https://github.com/losipiuk/presto/commit/7c8ffa09d4fe6c19480a210a352421fc56373d92